indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/InputSourceSampler.java

```diff
@@ -131,17 +130,27 @@ public SamplerResponse sample(
             continue;
           }
 
-          for (InputRow row : inputRowListPlusRawValues.getInputRows()) {
-            index.add(new SamplerInputRow(row, counter), true);
+          for (int i = 0; i < rawColumnsList.size(); i++) {
+            Map<String, Object> rawColumns = rawColumnsList.get(i);
+            InputRow row = inputRowListPlusRawValues.getInputRows().get(i);
+
+            //keep the index of the row to be added to responseRows for further use
+            final int rowIndex = responseRows.size();
+            index.add(new SamplerInputRow(row, rowIndex), true);
+
             // store the raw value; will be merged with the data from the IncrementalIndex later
-            responseRows[counter] = new SamplerResponseRow(rawColumns, null, null, null);
-            counter++;
+            responseRows.add(new SamplerResponseRow(rawColumns, null, null, null));
             numRowsIndexed++;
           }
         }
         catch (ParseException e) {
-          responseRows[counter] = new SamplerResponseRow(rawColumns, null, true, e.getMessage());
-          counter++;
+          if (rawColumnsList != null) {
+            responseRows.addAll(rawColumnsList.stream()
+                                              .map(rawColumns -> new SamplerResponseRow(rawColumns, null, true, e.getMessage()))
+                                              .collect(Collectors.toList()));
```

Comment:
Can this introduce duplicate rows in `responseRows`? Suppose you have 3 rawColumns in `rawColumnsList` including an unparseable row at the second position. The first row will be added to both `index` and thus a new `SamplerResponseRow` of the first `rawColumns` will be added to `responseRows`. But for the second row, `index.add()` will throw a `ParseException` which will execute these lines. In this case, 2 duplicate `rawColumns` of the first row will be added in `responseRows`.

Looking at [what `JsonReader` does](https://github.com/apache/druid/pull/10383/files#diff-c745a28591c789098f0a9efd75769ad242b6c40379d155fc61143ed63dccc995R93-R108), it seems throwing away the whole `intermediateRow` when there is any unparseable row. The sampler behavior should match to the actual ingestion.

Reply:
I notice that `index.add` does not throw `ParseException` but returns the exception in its returning result. The previous version of `InputSourceSampler` checks the result to rethrow this exception, while the [change](https://github.com/apache/druid/pull/10336/files#diff-a35198e960bc90b96d559e9ea93be1d4de52655e9a4f04016173114952ae0001) introduced by #10336 deletes these exception related code, which means the code here won't throw ParseException.

Of course, there's possibility that the code in for-loop throws `ParseException` in future. I've made little changes to avoid this potential situation to happen. Let's check it once the CI passes. 

BTW, my local branch has failed to build since last time it was rebased on master, I have to reply on the CI to check if there're any test case failures.

Reply:
> I notice that `index.add` does not throw `ParseException` but returns the exception in its returning result. The previous version of `InputSourceSampler` checks the result to rethrow this exception, while the [change](https://github.com/apache/druid/pull/10336/files#diff-a35198e960bc90b96d559e9ea93be1d4de52655e9a4f04016173114952ae0001) introduced by #10336 deletes these exception related code, which means the code here won't throw ParseException.
> 
> Of course, there's possibility that the code in for-loop throws `ParseException` in future. I've made little changes to avoid this potential situation to happen. Let's check it once the CI passes.

Oops, good point. I forgot what I have done in #10336.. I think `index.add()` should never throw parseException directly, but return them in `IncrementalIndexAddResult`. We can add this contract on the Javadoc of `IncrementalIndex.add()`. I suggest removing the `catch` clause here to avoid future confusion (there is no point in catching exceptions which cannot be thrown here).

Reply:
Got it. BTW, should we handle the exception returned in `IncrementalIndexAddResult` by adding the corresponding row to `responseRows` ?

Reply:
Yes, I think so because the sampler result should match to what will actually be ingested to Druid.

Reply:
A new case `testIndexParseException` in `InputSouceSamplerTest` has been added to check whether the exception returned by `index.add` has been processed properly. And also some little changes are applied in `IncrementalIndex.getCombinedParseException` and `SampleInputRow` to make exception message more accurate.